sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java

```diff
@@ -172,18 +177,30 @@ public static TableRow toTableRow(Row row) {
       Field schemaField = row.getSchema().getField(i);
       TypeName type = schemaField.getType().getTypeName();
 
-      if (TypeName.ARRAY == type) {
-        type = schemaField.getType().getCollectionElementType().getTypeName();
-        if (TypeName.ROW == type) {
-          List<Row> rows = (List<Row>) value;
-          List<TableRow> tableRows = new ArrayList<TableRow>(rows.size());
-          for (int j = 0; j < rows.size(); j++) {
-            tableRows.add(toTableRow(rows.get(j)));
+      switch (type) {
+        case ARRAY:
+          type = schemaField.getType().getCollectionElementType().getTypeName();
+          if (TypeName.ROW == type) {
+            List<Row> rows = (List<Row>) value;
+            List<TableRow> tableRows = new ArrayList<TableRow>(rows.size());
+            for (int j = 0; j < rows.size(); j++) {
+              tableRows.add(toTableRow(rows.get(j)));
+            }
+            value = tableRows;
           }
-          value = tableRows;
-        }
-      } else if (TypeName.ROW == type) {
-        value = toTableRow((Row) value);
+          break;
+        case ROW:
+          value = toTableRow((Row) value);
+          break;
+        case DATETIME:
+          DateTimeFormatter patternFormat = new DateTimeFormatterBuilder()
+              .appendPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZZ")
```

Comment:
The code that reads this wants something in the format `2016-06-29T22:45:46.123456+00:00`. I don't think this quite matches.

Reply:
Changing to `yyyy-MM-dd'T'HH:mm:ss.SSSSSSZZ`

Reply:
That's ISOFormat. That's why I linked to it. You don't have to have your own.

Reply:
(there's some other more recent RFC, which is compatible)

Reply:
The intention here is to send something that BigQuery expects because BigQuery can support `.SSSSSS`.

But after another thought, since `Joda` only supports milliseconds and ISO format, I will just stay with `"yyyy-MM-dd'T'HH:mm:ss.SSSZZ"` to be consistent with `Joda`

Reply:
OK. I see. This may be in a super tight loop and setting up the formatter in advance would save a decent amount of GC, and you could name it `BigQueryFormatter` to be clear why it exists. But OTOH it is pretty clear, so let's get the functionality in and be data-driven in our optimization.